File: python/triangle/triangle.py
def valid_triangle(f):
    return lambda s: all(s) and 2 * max(s) < sum(s) and f(s)


@valid_triangle
def equilateral(sides):
    return len(set(sides)) == 1

@valid_triangle
def isosceles(sides):
    return len(set(sides)) <= 2

@valid_triangle
def scalene(sides):
    return not isosceles(sides)

File: ruby/triangle/triangle.rb
class Triangle
  attr_reader :invalid_triangle
  def initialize(sides)
    @sides = sides
    @invalid_triangle = invalid_triangle
  end

  def equilateral?
    return false if @invalid_triangle
    @sides[0] == @sides[1] ? @sides[0] == @sides[2] : false
  end

  def isosceles?
    return false if @invalid_triangle
    @sides[0] == @sides[1] || @sides[1] == @sides[2] || @sides[0] == @sides[2]
  end

  def scalene?
    return false if @invalid_triangle
    !isosceles?
  end

  private

  def invalid_triangle
    return true if @sides.any? { |e| e <= 0 }
    return true if @sides[0] + @sides[1] < @sides[2]
    (@sides[0] - @sides[1]).abs > @sides[2]
  end

end

module BookKeeping
  VERSION = 1
end